filebeat/beater/filebeat.go

```diff
@@ -55,50 +51,62 @@ func (fb *Filebeat) Setup(b *beat.Beat) error {
 func (fb *Filebeat) Run(b *beat.Beat) error {
 
 	var err error
+	config := fb.config.Filebeat
 
-	// Init channels
-	fb.publisherChan = make(chan []*input.FileEvent, 1)
+	// Channel from harvesters to spooler
+	publisherChan := make(chan []*input.FileEvent, 1)
 
 	// Setup registrar to persist state
-	fb.registrar, err = crawler.NewRegistrar(fb.FbConfig.Filebeat.RegistryFile)
+	registrar, err := registrar.New(config.RegistryFile)
 	if err != nil {
 		logp.Err("Could not init registrar: %v", err)
 		return err
 	}
 
-	fb.crawler = &crawler.Crawler{
-		Registrar: fb.registrar,
-	}
+	// Publishes event to output
+	publisher := publish.New(config.PublishAsync,
+		publisherChan, registrar.Channel, b.Publisher.Connect())
 
-	// Load the previous log file locations now, for use in prospector
-	err = fb.registrar.LoadState()
+	// Init and Start spooler: Harvesters dump events into the spooler.
+	spooler, err := spooler.New(config, publisherChan)
 	if err != nil {
-		logp.Err("Error loading state: %v", err)
+		logp.Err("Could not init spooler: %v", err)
 		return err
 	}
 
-	// Init and Start spooler: Harvesters dump events into the spooler.
-	fb.spooler = NewSpooler(fb.FbConfig.Filebeat, fb.publisherChan)
-
+	crawler, err := crawler.New(registrar, spooler, config.Prospectors)
 	if err != nil {
-		logp.Err("Could not init spooler: %v", err)
+		logp.Err("Could not init crawler: %v", err)
 		return err
 	}
 
-	fb.registrar.Start()
-	fb.spooler.Start()
+	// The order of starting and stopping is important. Stopping is inverted to the starting order.
+	// The current order is: registrar, publisher, spooler, crawler
+	// That means, crawler is stopped first.
+
+	// Start the registrar
+	registrar.Start()
+	// Stopping registrar will write last state
+	defer registrar.Stop()
+
+	// Start publisher
+	publisher.Start()
+	// Stopping publisher (might potentially drop items)
+	defer publisher.Stop()
```

Comment:
We now have the flexibility to make optimizations here if needed. Currently the logic should be identical to what we had before.